pkg/ptr/ref.go

```diff
@@ -0,0 +1,137 @@
+package ptr
+
+import (
+	"cmp"
+	"database/sql"
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+
+	"github.com/BurntSushi/toml"
+)
+
+// Zeroable is a type constraint with types which supports zero references.
+// This includes int, bool, and string types.
+type Zeroable interface {
+	~bool | cmp.Ordered
+}
+
+// Ref is a reference type that holds a value of type T.
+// It provides methods to check if the reference is zero, get the value.
+// It is useful for working with optional values in a type-safe manner.
+type Ref[T Zeroable] struct {
+	v *T
+}
```

Comment:
For the record, my implementation is based on wrapping the value with pointer. The `go-optional` package uses a slice but that does not change a thing, it is actually worse - slice is neither comparable and ordered but struct is at least comparable (as long as it fields are comparable):

```
--- FAIL: TestUnmarshalYAMLMultipleDocuments (0.00s)
    --- FAIL: TestUnmarshalYAMLMultipleDocuments/simple (0.00s)
panic: ptr.Ref[string] is not a comparable Go type [recovered]
        panic: ptr.Ref[string] is not a comparable Go type
```